### PR TITLE
Refactor decision logic removing for loop

### PR DESCRIPTION
**Pull request by wxing1292**
* Remove workflowStartEventProvider in favor of mutableState GetStartEvent function
* Refactor handleDecisionTaskStarted to use history engine 
* Remove getWorkflowStartedEvent function

### CODE REVIEW

File: service/history/decisionHandler.go [hunk 1]
```diff
@@ -110,7 +109,10 @@ func (handler *decisionHandlerImpl) handleDecisionTaskScheduled(
 				transferTasks:  []persistence.Task{&persistence.RecordWorkflowStartedTask{}},
 			}
 
-			startEvent, _ := msBuilder.GetStartEvent()
+			startEvent, found := msBuilder.GetStartEvent()
+			if !found {
```
Comment (yux0) [inline service/history/decisionHandler.go]: Will this check break cron workflow?

File: service/history/decisionHandler.go [hunk 2]
```diff
@@ -295,17 +285,17 @@ func (handler *decisionHandlerImpl) handleDecisionTaskCompleted(
...
-	context, release, err0 := handler.historyCache.getOrCreateWorkflowExecutionWithTimeout(ctx, domainID, workflowExecution)
-	if err0 != nil {
-		return nil, err0
+	context, release, err := handler.historyCache.getOrCreateWorkflowExecutionWithTimeout(ctx, domainID, workflowExecution)
+	if err != nil {
+		return nil, err
 	}
 	defer func() { release(retError) }()
 
 Update_History_Loop:
```
Comment (yux0) [inline service/history/decisionHandler.go]: Could this also be replaced with updateWorkflowExecutionWithAction?

File: service/history/decisionHandler.go (hunk 1 above)
Comment (wxing1292) [reply service/history/decisionHandler.go]: these are missing checks. without this, code below will encounter nil pointer exception

File: service/history/decisionHandler.go (hunk 2 above)
Comment (wxing1292) [reply service/history/decisionHandler.go]: handling of decision task completed is really complex.
for now, i prefer stop here

**Pull request merged**